### ISSUE TITLE
Should implement UDP GSO on Linux

### ISSUE THREAD

**Issue by iriszero** [private-octopus/picoquic#1090]
Hi, I've been evaluating the performance of `picoquic` in the following environment.

Two machines, a client and server, each equipped with
* Intel(R) Xeon(R) CPU X5650; with AES-NI on
* Intel Corporation 82599ES 10-Gigabit SFI/SFP+ Network Connection
* Ubuntu 18.04.5 LTS
  * Kernel Params, with which `iperf3` yields 10Gbps for both TCP and UDP.
  ```
  sudo sysctl -w net.core.rmem_max=1342177280
  sudo sysctl -w net.core.wmem_max=516777216
  sudo sysctl -w net.core.rmem_default=10000000
  sudo sysctl -w net.core.wmem_default=10000000
  sudo sysctl -w net.core.netdev_max_backlog=416384
  sudo sysctl -w net.core.optmem_max=524288
  sudo sysctl -w net.ipv4.udp_mem=11416320 15221760 22832640
  sudo sysctl -w net.core.netdev_budget=1024
  ```
* DRAM 19GB  

and connected with a 10G link in the same rack.

With the args same as #78 
```
On server:
picoquicdemo -p 4433

On client:
picoquicdemo -D {server_ip} 4433 /200000000
```
The average of my three runs is 689.5 Mbps, which is slightly better than reported at https://github.com/private-octopus/picoquic/issues/78#issuecomment-484197079

My CPU profiling indicates that encryption accounts for 18.5% on the server and 11.0% on the client. ( SVG file attached: [cpu_prof.zip](https://github.com/private-octopus/picoquic/files/5521018/cpu_prof.zip) -> EDIT: wrong attachment for the server profiling; see the comment below. )

Meanwhile, `nginx`, with the same HW settings and HTTP2 + TLS on, shows 3,170 Mbps.

Despite the difference between QUIC and TCP+TLS layer, there seems large room for improving `picoquic`'s performance, near to that of `nginx`.

What can I try more? What should I look at?

**Comment by huitema** [comment]
@iriszero Thanks for the profiling. From previous work, I am aware of two issues: the socket code, and the unsophisticated nature of the "delayed ACK" algorithm. The profiling points to other possible areas, but I would like to solve the socket and ACK issues first.

The socket code is outside of the main stack. Each application tends to bring its own, for example because they rely on libevent. For the demo program (picoquicdemo), the unix code is in "sockloop.c". This is simple code using "select", "recvmsg" and "sendmsg". I started developing a faster code for the windows implementation in "winsockloop.c", which required adding a "prepare multiple packets" option in the API "picoquic_prepare_next_packet_ex()". The low hanging fruit is to develop a "linux" version of the loop, using that same API to take advantage of GSO, and/or io_uring. 

I plan to revisit the "delayed ACK" algorithm soon. This is tied to GSO. GSO sends multiple packets in a batch; if the batch generates multiple ACKs, the benefit will be lost.

Once the socket bottleneck are cleared, the profiling will probably show other areas for improvement. For example, the multiple packet API could bypass a lot of code that it currently does not. We could have a symmetric API for receiving data.

**Comment by huitema** [comment]
Something else. From looking at the flame diagrams, I realize that the cost of encryption on server is much more than 18.5%. What I see is:

Ops | Share
-- | --
_bsaes_encrypt8 | 15.41%
_bsaes_key_convert | 0.29%
_bsaes_ctr32_encrypt_blocks | 1.21%
gcm_ghash_4bit | 19.31%
gcm_stream_update | 0.14%
generic_aes_cipher_update | 0.14%
ossi_cipher_generic_einit | 6.00%
_x86_64_AES_encrypt_compact | 3.12%
gcm_get_ctx_param | 0.14%
gcm_stream_final | 0.06%
**Total crypto** | 45.82%

The cost of crypto appears way higher than what benchmarks of the openssl library predicts. Are you sure that the crypto library is compiled with the AES instructions?

**Comment by iriszero** [comment]
@huitema Thank you for the quick reply. As for the second comment, that's my mistake; a wrong attachment. Attaching the results of two runs on the server with _AES-NI on_:  [server-aesni-on.zip](https://github.com/private-octopus/picoquic/files/5521777/server-aesni-on.zip).

**Comment by iriszero** [comment]
As for the first comment, let me look into that later. Please feel free to close or keep open this issue. @huitema

**Comment by huitema** [comment]
Yes, the AES-NI diagram looks much more like what I saw before. 40% of time sending through sockets, 7.2% for select and recv_msg. Implementing GSO should alleviate that a lot.

**Comment by iriszero** [comment]
@huitema how much performance gain do you expect from GSO implementation?

**Comment by huitema** [comment]
If sendmsg is 40% of the current total, I assume that UDP GSO could reduce it to 10%. That means the max speed would go up from 700 Mbps to 1Gbps. At that point the next bottleneck will become obvious, and we will address that too, etc That way, we could plausibly get to 2 Gbps, maybe a bit more.

**Comment by huitema** [comment]
At some point, the crypto will come to dominate, and the solution will have to be "faster crypto". If crypto is about 20% at 700 Mbps, it would be 100% at 3.5 Gbps. But we need some CPU for the protocol. Or better crypto.

**Comment by iriszero** [comment]
@huitema when you said "I started developing a faster code for the windows implementation in "winsockloop.c"," it has NOT been implemented yet, thus not currently present in the `master` branch. Correct?

**Comment by huitema** [comment]
It is present on the master branch, but not compiled in the Linux builds, only in the windows builds. Turns out that the improvement of Windows' GSO equivalent are modest, because the UDP instructions are still executed on the main thread. To change that, I need to program the asynchronous Windows calls to use completion ports and a thread pool. I did not have the time to do that yet.

For the protocol stack proper, the main change is with the API:
```
int picoquic_prepare_next_packet_ex(picoquic_quic_t* quic, 
    uint64_t current_time, uint8_t* send_buffer, size_t send_buffer_max, size_t* send_length, 
    struct sockaddr_storage* p_addr_to, struct sockaddr_storage* p_addr_from, int* if_index,
    picoquic_connection_id_t* log_cid, picoquic_cnx_t** p_last_cnx, size_t* send_msg_size);
```
This API passes a "send_buffer" to the stack. If the buffer size `send_buffer_max` is larger than the size of one packet, if data is ready, and if flow, pacing and congestion control permit, the API may fill the buffer with several packets. The first packets will all have the same size, returned in `send_msg_size`.

The same API can be used to prepare packets for Linux GSO.

**Comment by iriszero** [comment]
May I ask one more thing, 
https://github.com/private-octopus/picoquic/issues/1022#issuecomment-681199976
> You cannot use UDP segmentation in QUIC -- spec says so. The main problem is what happens in case of packet loss: lose a single segment, you need to repeat the whole packet. That's why we are looking at variations of sendmmsg().

why is this the case?

[`draft-32`](https://tools.ietf.org/html/draft-ietf-quic-transport-32#section-12.2) says
> The receiver of coalesced QUIC packets MUST individually process each QUIC packet and separately acknowledge them, as if they were received as the payload of different UDP datagrams. For example, if decryption fails (because the keys are not available or any other reason), the receiver MAY either discard or buffer the packet for later processing and MUST attempt to process the remaining packets.

**Comment by huitema** [comment]
The coalesced packets are only used during the handshake. You cannot coalesce the 1RTT packets used for data transfer. Even if you could, the UDP reassembly process in the kernel would drop the entire UDP packet if a segment is missing.

**Comment by huitema** [comment]
@iriszero please check PR #1102 . I tuned the "ack frequency" to send fewer acks, thus create fewer interruptions. It has the side effect of taking better advantage of GSO. The perf on loopback on an Ubuntu VM is now 1.2 Gbps with BBR, 1.3 Gbps with Linux -- about 30% faster than before. Would be interesting to find the next bottleneck -- do I need more aggressive GSO, or something else.

**Comment by huitema** [comment]
Closing issue, as it is solved in PR #1102
